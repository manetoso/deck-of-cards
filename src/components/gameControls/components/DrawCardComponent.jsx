import { useEffect, useState } from 'react'

export function DrawCardComponent({ drawCard, gameFinished }) {
  const [buttonDisable, setButtonDisable] = useState(false)

  const handleButtonClick = async () => {
    if (buttonDisable) return

    // setButtonDisable(true)
    await drawCard()
  }

  useEffect(() => {
    if (buttonDisable) {
      const timeout = setTimeout(() => {
        setButtonDisable(false)
      }, 1000)

      // Clear the timeout if the component unmounts
      return () => clearTimeout(timeout)
    }
  }, [buttonDisable])
  return (
    <button className='btn' disabled={gameFinished | buttonDisable} onClick={handleButtonClick}>
      Draw a Card
    </button>
  )
}
