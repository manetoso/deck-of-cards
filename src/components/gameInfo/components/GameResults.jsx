export function GameResults({ results }) {
  return (
    <>
      <h3>Results:</h3>
      <small>
        *In the results only appear the cards that came out before the Queen of
        every suit*
      </small>
      <section className='game-results'>
        <p>
          <strong>Clubs:</strong> {results.clubs.join(', ')}
        </p>
        <p>
          <strong>Diamonds:</strong> {results.diamonds.join(', ')}
        </p>
        <p>
          <strong>Hearts:</strong> {results.hearts.join(', ')}
        </p>
        <p>
          <strong>Spades:</strong> {results.spades.join(', ')}
        </p>
      </section>
    </>
  )
}
