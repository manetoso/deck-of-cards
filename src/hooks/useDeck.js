import { useState, useCallback, useEffect } from 'react'
import { deckApi, backOfCardImage } from '../api/deckApi'

/**
 * QC = Queen of Clubs
 * QD = Queen of Diamonds
 * QH = Queen of Hearts
 * QS = Queen of Spades
 */

const resultsInitialState = {
  clubs: [],
  diamonds: [],
  hearts: [],
  spades: []
}

export const useDeck = () => {
  const [currentDeck, setCurrentDeck] = useState({
    deck_id: '',
    remaining: 0,
    shuffled: false
  })
  const [playerPile, setPlayerPile] = useState([])
  const [queenCounter, setQueenCounter] = useState(0)
  const [gameFinished, setGameFinished] = useState(false)
  const [results, setResults] = useState(resultsInitialState)

  const createDeck = useCallback(async () => {
    try {
      const { data } = await deckApi.get('/new/shuffle/?deck_count=1')
      const { success, ...rest } = data
      setCurrentDeck(rest)
    } catch (error) {
      console.error(error)
    }
  }, [])

  const drawCard = async () => {
    if (currentDeck.remaining === 0) return alert('No more cards in deck')
    try {
      const { data } = await deckApi.get(
        `/${currentDeck.deck_id}/draw/?count=1`
      )
      const { cards, remaining } = data
      const {
        code,
        value,
        images: { svg }
      } = cards[0]
      setPlayerPile([...playerPile, { code, value, img: svg }])
      setCurrentDeck({ ...currentDeck, remaining })
      if (code.includes('Q')) setQueenCounter(queenCounter + 1)
    } catch (error) {
      console.error(error)
    }
  }

  const resetGame = async () => {
    await createDeck()
    setGameFinished(false)
    setResults(resultsInitialState)
    setPlayerPile([])
    setQueenCounter(0)
  }

  const searchQuenInPile = (pile, suit) => {
    const pileOdSuit = pile.filter((card) => card.code.includes(suit))
    const queenIndex = pileOdSuit.findIndex((card) => card.code.includes('Q'))
    const newPile = pileOdSuit.slice(0, queenIndex + 1)

    const compareCardValues = (a, b) => {
      const cardValues = 'A1234567890JQK' // Define the order of card values
      const aValue = cardValues.indexOf(a.code.charAt(0))
      const bValue = cardValues.indexOf(b.code.charAt(0))
      return aValue - bValue
    }
    newPile.sort(compareCardValues)
    return newPile
  }

  const finishGame = () => {
    setGameFinished(true)
    alert('Game Finished!')
    const clubsPile = searchQuenInPile(playerPile, 'C')
    const diamondsPile = searchQuenInPile(playerPile, 'D')
    const heartsPile = searchQuenInPile(playerPile, 'H')
    const spadesPile = searchQuenInPile(playerPile, 'S')
    const clubs = clubsPile.map((card) => card.value)
    const diamonds = diamondsPile.map((card) => card.value)
    const hearts = heartsPile.map((card) => card.value)
    const spades = spadesPile.map((card) => card.value)
    setResults({
      clubs,
      diamonds,
      hearts,
      spades
    })
  }

  useEffect(() => {
    if (queenCounter === 4) {
      finishGame()
    }
  }, [queenCounter])

  useEffect(() => {
    createDeck()
  }, [createDeck])

  return {
    backOfCardImage,
    currentDeck,
    drawCard,
    gameFinished,
    playerPile,
    queenCounter,
    resetGame,
    results
  }
}
