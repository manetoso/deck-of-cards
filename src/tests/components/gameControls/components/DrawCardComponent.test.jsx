import {
  act,
  render,
  renderHook,
  screen,
  waitFor
} from '@testing-library/react'

import { useDeck } from '../../../../hooks/useDeck'
import { DrawCardComponent } from '../../../../components/gameControls/components/DrawCardComponent'

describe('DrawCardComponent tests', () => {
  test('DrawCardComponent renders correctly', () => {
    const { container } = render(<DrawCardComponent />)
    // create a new snapshot file
    expect(container).toMatchSnapshot()
  })

  test('should be disable for a second once its called', async () => {
    const { result } = renderHook(() => useDeck())

    await waitFor(() => {})

    const { drawCard, gameFinished } = result.current
    render(
      <DrawCardComponent drawCard={drawCard} gameFinished={gameFinished} />
    )
    screen.debug()
    const button = screen.getByRole('button')
    act(() => {
      button.click()
    })
    expect(button.disabled).toBe(true)
    await waitFor(
      () => {
        expect(button.disabled).toBe(false)
      },
      { timeout: 1100 }
    )
  })
})
