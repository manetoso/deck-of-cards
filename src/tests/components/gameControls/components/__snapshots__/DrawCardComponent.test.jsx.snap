// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawCardComponent tests DrawCardComponent renders correctly 1`] = `
<div>
  <button
    class="btn"
  >
    Draw a Card
  </button>
</div>
`;
